File: frontend/src/lib/api/agent-keys.ts
import apiClient from './client'

/** Agent-key permission scopes, least→most privileged (read < write < admin). */
export type AgentPermission = 'read' | 'write' | 'admin'

/** A key as listed back — secret-free (prefix + metadata, never the plaintext). */
export interface AgentKey {
  id: string
  agent_id: string
  key_prefix: string
  label?: string | null
  permission: AgentPermission
  rate_limit_rpm?: number | null
  revoked: boolean
  created?: string | null
  last_used_at?: string | null
}

/** The mint response — carries the plaintext key EXACTLY ONCE. */
export interface AgentKeyCreated extends AgentKey {
  key: string
}

export interface MintAgentKeyInput {
  agent_id: string
  permission: AgentPermission
  label?: string
  rate_limit_rpm?: number
}

/** One audit-log row for an agent's API calls. */
export interface AgentAuditEntry {
  agent_id?: string
  method?: string
  path?: string
  status?: number
  latency_ms?: number | null
  job_id?: string | null
  ts?: string | null
}

export interface KeyAuditLog {
  key_id: string
  agent_id: string
  entries: AgentAuditEntry[]
}

/**
 * Session-authed agent-key management (Track G.4). These call the operator
 * surface under `/api/agent-keys` (shared-password protected), NOT the
 * `X-API-Key` agent capability routes.
 */
export const agentKeysApi = {
  list: async (): Promise<AgentKey[]> => {
    const res = await apiClient.get<AgentKey[]>('/agent-keys')
    return res.data
  },

  mint: async (input: MintAgentKeyInput): Promise<AgentKeyCreated> => {
    const res = await apiClient.post<AgentKeyCreated>('/agent-keys', input)
    return res.data
  },

  revoke: async (id: string): Promise<void> => {
    await apiClient.delete(`/agent-keys/${id}`)
  },

  auditLog: async (id: string, limit = 100): Promise<KeyAuditLog> => {
    const res = await apiClient.get<KeyAuditLog>(`/agent-keys/${id}/audit-log`, {
      params: { limit },
    })
    return res.data
  },
}

File: apps/app-main/src/app_main/api/routers/agent_keys.py
"""Agent-key MANAGEMENT endpoints (Track G.1).

Human-operator surface for minting / listing / revoking agent keys. These live
under ``/api`` and are therefore protected by the existing shared-password
middleware (session auth) — NOT by ``X-API-Key``. Agents authenticate to the
capability routes (``/api/v1/agents/*``); people manage keys here.

The plaintext key is returned EXACTLY ONCE from the mint response and never again
(only its SHA-256 hash is stored). Listings expose ``key_prefix`` + metadata only.
"""

from fastapi import APIRouter, Depends, HTTPException
from loguru import logger
from shared.models.agents import AgentKeyCreate, AgentKeyCreated, AgentKeyPublic

from app_main.services.agents.audit_service import AgentAuditService
from app_main.services.agents.key_service import AgentKeyService

router = APIRouter(prefix="/agent-keys", tags=["agent-keys"])


def get_agent_key_service() -> AgentKeyService:
    return AgentKeyService()


@router.post("", response_model=AgentKeyCreated, status_code=201)
async def mint_agent_key(
    body: AgentKeyCreate,
    service: AgentKeyService = Depends(get_agent_key_service),
) -> AgentKeyCreated:
    """Mint a new agent key. The plaintext is returned here ONCE — store it now."""
    if body.permission not in ("read", "write", "admin"):
        raise HTTPException(status_code=422, detail="invalid permission")
    try:
        public, plaintext = await service.mint(
            agent_id=body.agent_id,
            permission=body.permission,
            label=body.label,
            rate_limit_rpm=body.rate_limit_rpm,
        )
    except Exception as e:  # noqa: BLE001
        logger.error(f"Minting agent key failed: {e}")
        raise HTTPException(status_code=500, detail=f"Could not mint key: {str(e)}")
    return AgentKeyCreated(key=plaintext, **public)


@router.get("", response_model=list[AgentKeyPublic])
async def list_agent_keys(
    service: AgentKeyService = Depends(get_agent_key_service),
) -> list[AgentKeyPublic]:
    """List all agent keys (secret-free: prefix + metadata, never the plaintext)."""
    rows = await service.list()
    return [AgentKeyPublic(**r) for r in rows]


@router.get("/{key_id}/audit-log")
async def get_key_audit_log(
    key_id: str,
    limit: int = 100,
    service: AgentKeyService = Depends(get_agent_key_service),
) -> dict:
    """A key's recent API calls (session-authed operator view, G.4).

    The agent capability router exposes the audit-log only behind ``X-API-Key``
    (which the operator doesn't hold); this session-authed view resolves the key
    to its ``agent_id`` server-side and returns that agent's trail. Note the trail
    is keyed by ``agent_id`` — keys that SHARE an agent_id share the trail.
    """
    key = await service.get(key_id)
    if key is None:
        raise HTTPException(status_code=404, detail="key not found")
    limit = max(1, min(limit, 500))
    entries = await AgentAuditService().list_for_agent(key["agent_id"], limit=limit)
    return {"key_id": key_id, "agent_id": key["agent_id"], "entries": entries}


@router.delete("/{key_id:path}", status_code=204)
async def revoke_agent_key(
    key_id: str,
    service: AgentKeyService = Depends(get_agent_key_service),
) -> None:
    """Revoke a key by id (idempotent)."""
    await service.revoke(key_id)
